src/compiler/checker.ts

```diff
@@ -3874,28 +3910,25 @@ namespace ts {
             return links.resolvedType;
         }
 
-        function addTypeToSortedSet(sortedSet: Type[], type: Type) {
-            if (type.flags & TypeFlags.Union) {
-                addTypesToSortedSet(sortedSet, (<UnionType>type).types);
+        function addTypeToSet(typeSet: Type[], type: Type, typeKind: TypeFlags) {
+            if (type.flags & typeKind) {

```

Comment:
Oh, then was it wrong to do it here? https://github.com/Microsoft/TypeScript/blob/master/src/compiler/checker.ts#L410